src/vsphere_cpi/spec/unit/cloud/vsphere/vm_placement_selection_pipeline_spec.rb

```diff
@@ -0,0 +1,135 @@
+require 'spec_helper'
+require 'cloud/vsphere/disk_placement_selection_pipeline'
+require 'cloud/vsphere/resources/datastore'
+
+describe VSphereCloud::VmPlacementSelectionPipeline do
+  subject { described_class.new(*criteria) {[placement_1, placement_2]} }
+
+  def fake_placement(cluster)
+    VSphereCloud::VmPlacement.new(cluster: cluster, datastores: cluster.accessible_datastores.values, hosts: nil)
+  end
+
+
+  def fake_datastore(name, free_space: 4096)
+    VSphereCloud::Resources::Datastore.new(
+      name, true, instance_double('VimSdk::Vim::Datastore'), 8192, free_space
+    ).tap do |resource|
+      allow(resource).to receive(:accessible?).and_return(true)
+    end
+  end
+
+  def fake_cluster(name, *args)
+    instance_double('VSphereCloud::Resources::Cluster',
+      :name => name,
+      :free_memory => 8192,
+      :accessible_datastores => args.map{|ds| [ds.name, ds]}.to_h )
+  end
+
+  let(:criteria) { [disk_config: disk_config, req_memory: 1024] }
+  let(:ds_cl1_1) { fake_datastore('fake-ds-cl1-1') }
+  let(:ds_cl1_2) { fake_datastore('fake-ds-cl1-2') }
+  let(:ds_cl2_1) { fake_datastore('fake-ds-cl2-1') }
+  let(:ds_cl2_2) { fake_datastore('fake-ds-cl2-2') }
+  let(:ds_shared) { fake_datastore('fake-ds-shared') }
+  let(:cluster_1) { fake_cluster('fake-cluster-1', ds_cl1_1, ds_cl1_2, ds_shared) }
+  let(:cluster_2) { fake_cluster('fake-cluster-2', ds_cl2_1, ds_cl2_2, ds_shared) }
+  let(:placement_1) { fake_placement(cluster_1) }
+  let(:placement_2) { fake_placement(cluster_2) }
+  let(:disk_config) {[]}
+
+  context 'when there are no disks to be configured' do
+    it 'only generates placements with enough memory' do
+      allow_any_instance_of(VSphereCloud::Resources::Datastore).to receive(:maintenance_mode?).and_return(false)
```

Comment:
I would rather keep it as a single context then. maintenance-mode? of datastores is just a filter and making a different context for it would be confusing.
Right now I have arranged it in a way where we either configure 0 disks and concentrate on all other filters or have 3 disks.